### PR TITLE
unix,doc: always copy process title when setting

### PR DESCRIPTION
**Pull request by 64**
Ensures that the user's `argv` is copied into a local buffer when calling `uv_setup_args`. Before, the `argv` was simply being pointed to, which meant that libuv could end up accessing invalid memory if the user decided to later edit the memory at that location. It also meant that a subsequent call to `uv_set_process_title` would never write more characters than the length of `argv[0]`.

With the new changes, `argv[0]` is copied into a temporary buffer and any subsequent calls to `uv_set_process_title` will thus be able to copy as many characters as the call to `uv__strdup` permits. Note that on *BSD and AIX this behaviour was already in effect .

Some error checking (specifically checking the result of `uv__strdup`) has been added, and calls to `uv__free` rearranged so that in case of `ENOMEM` `uv__free` can't be called erroneously.

Finally, a warning about the thread safety of `uv_get_process_title` and `uv_set_process_title` was added to the docs.

Note: I wasn't too sure how to add a test for this, since I don't believe the current test system exposes `argv`.

Addresses https://github.com/libuv/libuv/issues/1395.

### CODE REVIEW

**Comment by cjihrig** [comment]
I haven't reviewed the code, but can you split this into two commits - one for the docs, and one for the code changes.

**Comment by 64** [comment]
> I haven't reviewed the code, but can you split this into two commits - one for the docs, and one for the code changes.

Done.

**Comment by 64** [comment]
bump.

**Comment by saghul** [comment]
@cjihrig can you PTAL too?

File: src/unix/proctitle.c [hunk 1]
```diff
@@ -50,10 +50,14 @@ char** uv_setup_args(int argc, char** argv) {
...
-  process_title.len = strlen(argv[0]);
+  process_title.str = uv__strdup(argv[0]);
+  if (process_title.str == NULL)
+    return argv;
+  process_title.len = strlen(process_title.str);
 #else
-  process_title.str = argv[0];
+  process_title.str = uv__strdup(argv[0]);
+  if (process_title.str == NULL)
+    return argv;
```
Comment (bnoordhuis) [inline src/unix/proctitle.c]: Can you hoist the common code (everything up to and including the if statement) out of the `#if` block?

**Comment by 64** [comment]
Comment addressed.

**Comment by bnoordhuis** [comment]
https://ci.nodejs.org/job/libuv-test-commit-linux/357/nodes=armv7-wheezy/console - failures look unrelated, seems like a stray process hogging a port.

**Comment by cjihrig** [comment]
Landed in https://github.com/libuv/libuv/commit/bdc87005a17ce843e0efca8938aa8c190aa91a38 and https://github.com/libuv/libuv/commit/78c17238f48d9083359206a2215fc63dd7a0283d. Thanks!

**Comment by cjihrig** [comment]
This causes a test failure in Node. See https://github.com/nodejs/node/pull/14866.

The issue appears to be that overwriting `argv[0]` is no longer possible. So, the process title changes within the process, but externally, the original `argv[0]` is seen.

**Comment by 64** [comment]
I don't see `argv[0]` being overwritten, where are you looking? Unless I'm misunderstanding you.

**Comment by cjihrig** [comment]
So, the original post said:

> Before, the argv was simply being pointed to, which meant that libuv could end up accessing invalid memory if the user decided to later edit the memory at that location.

and

> With the new changes, argv[0] is copied into a temporary buffer and any subsequent calls to uv_set_process_title will thus be able to copy as many characters as the call to uv__strdup permits. Note that on *BSD and AIX this behaviour was already in effect .

My understanding is that with the changes in this PR, we are no longer pointing to `argv`. So, when `uv_set_process_title()` is called, we are updating a local buffer, not `argv`. And, when `uv_get_process_title()` is called, we'd be reading from that buffer. However, outside of the process, tools like `ps` would still report the original `argv` values. Is that correct? If so, that explains the test failure in the Node.js test suite.

That also leaves the question of what to do about it.

**Comment by 64** [comment]
You're right about the temporary buffer, but I don't believe that changing `argv[0]` actually changes the name of the process title. In `unix/linux-core.c:uv__set_process_title` the function `prctl` is called which I believe does the switch.

**Comment by cjihrig** [comment]
Changing `argv[0]` [does change the name of the process name](https://unix.stackexchange.com/questions/167490/how-can-a-process-appear-to-have-different-name-in-ps-output). I think the difference between that and `prctl` is explained [here](https://stackoverflow.com/questions/16165667/change-process-name-without-changing-argv0-in-linux).

**Comment by 64** [comment]
My mistake then. So `argv[0]` needs to be changed too? We need to be careful about how much space is allocated for it, in that case. Is this only an issue on Linux?

**Comment by cjihrig** [comment]
> Is this only an issue on Linux?

I don't think so. The Node CI lit up bright red when I updated the version of libuv. You can see the failures here: https://ci.nodejs.org/job/node-test-commit/11826/

**Comment by 64** [comment]
Looks like you're right, although there are some unrelated Windows failures there too.

I think just reverting the changes in `unix/proctitle.c` from this PR would do the trick. I'll try later and see if that works on my machine. Might also be a good idea to improve the tests in libuv to check that the title is actually being set correctly, but I'm not sure of a good method of doing that in a portable way.

**Comment by cjihrig** [comment]
There are some things that are significantly harder or more involved to test in C than they are from Node. A CI job or something similar to test libuv in Node before a release is cut might be nice, but are beyond the scope of this issue.

**Comment by santigimeno** [comment]
> A CI job or something similar to test libuv in Node before a release is cut might be nice

That's indeed a very good idea. /cc @refack as I think it's now member of the nodejs build group

**Comment by refack** [comment]
`it's`? 😱

**Comment by santigimeno** [comment]
OMG sorry :(. I'll edit the comment.

**Comment by refack** [comment]
> CI job or something similar to test libuv in Node before a release is cut might be nice, but are beyond the scope of this issue.

~I'm not sure what would that entail?~ Ohh like a lib CitGM... that's a nice idea!

**Comment by refack** [comment]
> OMG sorry :(. I'll edit the comment.

P.S. I'm almost un-offendable

**Comment by cjihrig** [comment]
I was thinking more like `nodejs/node-v8` then CITGM. Let's move this to a separate issue though.

**Pull request closed without merging**